### PR TITLE
SSSE3 implementation of CfL predictor

### PR DESCRIPTION
**Pull request by barrbrain**
Not quite as fast as the C intrinsics implementation in libaom yet.
Already a lot faster than the current native rust implementation.

### CODE REVIEW

File: src/predict.rs [hunk 1]
```diff
@@ -426,11 +431,51 @@ pub trait Intra: Dim {
...
+    let alpha_sign = _mm_set1_epi16(alpha);
+    let alpha_q12 = _mm_slli_epi16(_mm_abs_epi16(alpha_sign), 9);
+    let dc_q0 = _mm_set1_epi16(output[0] as i16);
+    let max = _mm_set1_epi16((1 << bit_depth) - 1);
+
+    for (line, luma) in
+      output.chunks_mut(stride).take(Self::H).zip(ac.chunks(32))
+    {
+      let mut i = 0isize;
+      while (i as usize) < Self::W {
```
Comment (lu-zero) [inline src/predict.rs]: What happens if you unroll it 2x? (the fact that using the iterator does not unroll it might or might not be a bug.

Comment (barrbrain) [reply src/predict.rs]: This already fully unrolls for 4x4. I'll update this once more before we merge. Using pointer arithmetic for the outer loop eliminates branching.

File: src/predict.rs [hunk 2]
```diff
@@ -469,11 +470,13 @@ pub trait Intra: Dim {
     bit_depth: usize
   ) {
     if alpha == 0 { return; }
+    assert!(32 >= Self::W);
+    assert!(ac.len() >= 32 * Self::H);
+    assert!(stride >= Self::W);
+    assert!(output.len() >= stride * Self::H);
```
Comment (lu-zero) [inline src/predict.rs]: If you move them in the function body, does the compiler elide the checks for you even if you use the normal iterators/array indexing? (It should)

Comment (barrbrain) [reply src/predict.rs]: I'll try that for the next change. The CI failures highlight that these assertions are imprecise. I'll update to permit CfL on the bottom row in positions other than the left-most column.

**Pull request merged**